relay-metrics/src/aggregator.rs

```diff
@@ -16,6 +17,278 @@ use tokio::time::Instant;
 use crate::bucket::{Bucket, BucketValue};
 use crate::protocol::{self, MetricNamespace, MetricResourceIdentifier};
 use crate::statsd::{MetricCounters, MetricGauges, MetricHistograms, MetricSets, MetricTimers};
+use crate::{
+    aggregator, Condition, DistributionValue, Field, HashedBucket, ScopedAggregatorConfig,
+};
+
+/// The fraction of [`AggregatorServiceConfig::max_flush_bytes`](crate::AggregatorServiceConfig) at which buckets will be split. A value of
+/// `2` means that all buckets smaller than half of max_flush_bytes will be moved in their entirety,
+/// and buckets larger will be split up.
+const BUCKET_SPLIT_FACTOR: usize = 32;
+
+/// The average size of values when serialized.
+const AVG_VALUE_SIZE: usize = 8;
+
+/// Static overhead used when calculating approximate serialized size of a bucket without value.
+const STATIC_OVERHEAD_SIZE: usize = 50;
+
+/// Utility that routes metrics to the appropriate aggregator.
+///
+/// Each aggregator gets its own configuration.
+/// Metrics are routed to the first aggregator which matches the configuration's [`Condition`].
+/// If no condition matches, the metrics are routed to the `default_aggregator`.
+pub struct AggregatorRouter {
+    default_aggregator: aggregator::Aggregator,
+    secondary_aggregators: BTreeMap<MetricNamespace, aggregator::Aggregator>,
+}
+
+impl AggregatorRouter {
+    /// Creates a new instance of [`AggregatorRouter`].
+    pub fn new(
+        aggregator_config: AggregatorConfig,
+        secondary_aggregators: Vec<ScopedAggregatorConfig>,
+    ) -> Self {
+        Self {
+            default_aggregator: aggregator::Aggregator::new(aggregator_config),
+            secondary_aggregators: secondary_aggregators
+                .into_iter()
+                .map(|c| {
+                    let Condition::Eq(Field::Namespace(namespace)) = c.condition;
+                    (
+                        namespace,
+                        aggregator::Aggregator::named(c.name, c.config.aggregator),
+                    )
+                })
+                .collect(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn default_bucket_count(&self) -> usize {
+        self.default_aggregator.bucket_count()
+    }
+
+    /// Returns an iterator of references to all the aggregators in the [`AggregatorRouter`].
+    fn aggregators_ref(&self) -> impl Iterator<Item = &aggregator::Aggregator> {
```

Comment:
nit: I would just call this `aggregators()`. See for example `values()`, `values_mut()` on `HashMap`: https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.values

```suggestion
    fn aggregators(&self) -> impl Iterator<Item = &aggregator::Aggregator> {
```